File: fluentify_frontend/lib/features/auth/domain/entities/user_entity.dart
import 'package:equatable/equatable.dart';

class UserEntity extends Equatable {
  final String id;
  final String email;
  final String fullName;
  final String? profileImage;
  final int coins;
  final int streakCount;
  final String level;
  final int missionsCompleted;
  final int gameLevel;
  final int grammarLevel;
  final int speakingLevel;
  final String subscriptionPlan;
  final DateTime? subscriptionExpiry;

  const UserEntity({
    required this.id,
    required this.email,
    required this.fullName,
    this.profileImage,
    required this.coins,
    required this.streakCount,
    required this.level,
    this.missionsCompleted = 0,
    this.gameLevel = 1,
    this.grammarLevel = 1,
    this.speakingLevel = 1,
    this.wordMatchLevel = 1,
    this.typingLevel = 1,
    this.dictationLevel = 1,
    this.readingLevel = 1,
    this.rapidFireLevel = 1,
    this.subscriptionPlan = 'free',
    this.subscriptionExpiry,
  });

  final int wordMatchLevel;
  final int typingLevel;
  final int dictationLevel;
  final int readingLevel;
  final int rapidFireLevel;

  bool get isPremium => subscriptionPlan == 'premium';

  @override
  List<Object?> get props => [
        id,
        email,
        fullName,
        profileImage,
        coins,
        streakCount,
        level,
        missionsCompleted,
        gameLevel,
        grammarLevel,
        speakingLevel,
        wordMatchLevel,
        typingLevel,
        dictationLevel,
        readingLevel,
        rapidFireLevel,
        subscriptionPlan,
        subscriptionExpiry,
      ];
}

File: fluentify_backend/src/modules/user/entities/user.entity.ts
import { Entity, Column, PrimaryGeneratedColumn, CreateDateColumn, UpdateDateColumn } from 'typeorm';

export enum UserRole {
    USER = 'user',
    MENTOR = 'mentor',
    ADMIN = 'admin',
}

export enum SubscriptionPlan {
    FREE = 'free',
    STARTER = 'starter',
    STANDARD = 'standard',
    PRO = 'pro',
    ELITE = 'elite',
    PREMIUM = 'premium', // Legacy support
    BASIC = 'basic', // Legacy support
}

@Entity('users')
export class User {
    @PrimaryGeneratedColumn('uuid')
    id: string;

    @Column({ unique: true })
    email: string;

    @Column({ nullable: true })
    password?: string;

    @Column()
    fullName: string;

    @Column({ nullable: true })
    avatarUrl?: string;

    @Column({
        type: 'enum',
        enum: UserRole,
        default: UserRole.USER,
    })
    role: UserRole;

    @Column({
        type: 'enum',
        enum: SubscriptionPlan,
        default: SubscriptionPlan.FREE,
    })
    subscriptionPlan: SubscriptionPlan;

    @Column({ type: 'timestamp', nullable: true })
    subscriptionExpiry?: Date;

    @Column({ default: 'A1' })
    cefrLevel: string; // A1, A2, B1, B2, C1, C2

    @Column({ default: 0 })
    coins: number;

    @Column({ default: 0 })
    streakCount: number;

    @Column({ default: 1 })
    gameLevel: number;

    @Column({ default: 1 })
    grammarLevel: number;

    @Column({ default: 1 })
    speakingLevel: number;

    @Column({ type: 'timestamp', nullable: true })
    lastPracticeDate?: Date;

    @Column({ default: true })
    isActive: boolean;

    @Column({ nullable: true })
    firebaseUid?: string;

    @CreateDateColumn()
    createdAt: Date;

    @UpdateDateColumn()
    updatedAt: Date;
}
